src/datasources/ipc.py

```diff
@@ -0,0 +1,231 @@
+    )
+    params = {
+        "app_identifier": os.getenv("HAPI_APP_IDENTIFIER"),
+        "admin_level": 0,
+        "output_format": "json",
+        "limit": 10000,
+        "offset": 0,
+    }
+    # Check if the request was successful
+    response = requests.get(endpoint, params=params)
+
+    json_data = response.json()
+    # Extract the data list from the JSON
+    data_list = json_data.get("data", [])
+    df_response = pd.DataFrame(data_list)
+
+    df_response["From"] = pd.to_datetime(df_response["reference_period_start"])
+    df_response["To"] = pd.to_datetime(df_response["reference_period_end"])
+    df_response["year"] = df_response["To"].dt.year
+    if filter_iso3s:
+        df_response = df_response[df_response.location_code.isin(filter_iso3s)]
+    return df_response.sort_values("reference_period_start", ascending=False)[
+        [
+            "location_code",
+            "ipc_phase",
+            "ipc_type",
+            "population_fraction_in_phase",
+            "population_in_phase",
+            "From",
+            "To",
+            "year",
+        ]
+    ]
+
+
+def classify_reports(df):
+    _df = df.copy()
+    _df = _combine_4_plus(_df)
+    _df = _get_pop_analyzed(_df)
+    _df = _transform_wide(_df)
+    _df["cat"] = _df.apply(_classify_row, axis=1)
+    split_categories = _df["cat"].str.split(" - ", expand=True)
+    _df["alert_level"] = split_categories[0]
+    _df["alert_level_detail"] = split_categories[1]
+    _df = _df.drop("cat", axis="columns")
+    return _df
+
+
+def _get_pop_analyzed(df):
+    _df = df.copy()
+    pop_all = _df[_df["ipc_phase"] == "all"].set_index(
+        ["location_code", "ipc_type", "From", "To"]
+    )["population_in_phase"]
+    _df["population_analyzed"] = (
+        _df.set_index(["location_code", "ipc_type", "From", "To"])
+        .index.map(pop_all)
+        .values
+    )
+    return _df
+
+
+def _combine_4_plus(df):
+    _df = df.copy()
+    mapping = {"4": "4+", "5": "4+"}
+    _df["ipc_phase"] = df["ipc_phase"].map(mapping).fillna(_df["ipc_phase"])
+    dff = _df[_df.ipc_phase == "4+"]
+    dff = (
+        dff.groupby(
+            [
+                "From",
+                "To",
+                "location_code",
+                "ipc_type",
+                "year",
+                "ipc_phase",
+            ]
+        )
+        .agg(
+            {
+                "population_fraction_in_phase": "sum",
+                "population_in_phase": "sum",
+            }
+        )
+        .reset_index()
+    )
+    return pd.concat([df, dff])
+
+
+def _classify_row(row):
+    P3 = row.get("proportion_3+", np.nan)
+    PP4 = row.get("population_4+", np.nan)
+    D3 = row.get("pt_change_3+", np.nan)
+    D4 = row.get("pt_change_4+", np.nan)
+
+    VH_S = PP4 >= VH_S_POP_4
+    VH_D = P3 >= VH_D_PR_3 and D4 >= VH_D_IN_4 * 100
+    H_S = PP4 >= H_S_POP_4
+    H_D = P3 >= H_D_PR_3 and D3 >= H_D_IN_3 * 100
+    M = (P3 >= M_PR_3) or (PP4 >= M_POP_4)
```

Comment:
if the underlying constants are renamed then perhaps these new objects canb e too